### ISSUE TITLE
Keep only model with best validation score

### ISSUE THREAD

**Issue by abuzarmahmood** [pulakatlab/auto_slide#12]

**Comment by abuzarmahmood** [comment]
The issue at hand involves modifying the training script to ensure that only the model with the best validation score is saved. Here's a summary of the suggestions and changes made by the agents:

1. **File to Modify:**
   - **File**: `/home/exouser/Desktop/blech_github_bot/src/repos/pulakatlab/auto_slide/src/pipeline/model/training.py`
   - **Description**: This file contains the training loop for a Mask R-CNN model using PyTorch. It includes logic for training, validation, and saving the model and loss histories.

2. **Plan to Address the Issue:**
   - Introduce a variable to track the best validation loss observed so far.
   - Update this variable whenever a new best validation loss is found.
   - Save the model only when the current validation loss is lower than the best validation loss.

3. **Specific Changes to be Made:**

   - **Track Best Validation Loss:**
     - **Line**: 155 (after initializing `all_val_losses`)
     - **Description**: Introduce a variable `best_val_loss` to track the best validation loss.
     - **Code Snippet**:
       ```python
       best_val_loss = float('inf')
       ```

   - **Update Best Validation Loss and Save Model:**
     - **Line**: 191-195 (within the validation loop)
     - **Description**: Update the logic to save the model only if the current validation loss is better than the best validation loss.
     - **Code Snippet**:
       ```python
       if val_epoch_loss < best_val_loss:
           best_val_loss = val_epoch_loss
           torch.save(model.state_dict(), os.path.join(artifacts_dir, 'mask_rcnn_model.pth'))
           print(f"New best model saved with validation loss: {best_val_loss}")
       ```

These changes will ensure that only the model with the best validation score is saved during training.

---
*This response was automatically generated by blech_bot using model gpt-4o*

---
*This response was automatically generated by blech_bot*

**Comment by abuzarmahmood** [comment]
[ develop_issue ]

**Comment by abuzarmahmood** [comment]
Created pull request: https://github.com/pulakatlab/auto_slide/pull/14
Continue discussion there.

---
*This response was automatically generated by blech_bot*
